openmc/arithmetic.py

```diff
@@ -0,0 +1,842 @@
+            string += '(' + str(self.left_nuclide)
+
+        string += ' ' + self.binary_op + ' '
+
+        # If the Summary was linked, the right nuclide is a Nuclide object
+        if isinstance(self.right_nuclide, openmc.Nuclide):
+            string += self.right_nuclide.name + ')'
+        # If the Summary was not linked, the right nuclide is the ZAID
+        else:
+            string += str(self.right_nuclide) + ')'
+
+        return string
+
+    @left_nuclide.setter
+    def left_nuclide(self, left_nuclide):
+        cv.check_type('left_nuclide', left_nuclide,
+                      (openmc.Nuclide, CrossNuclide, AggregateNuclide))
+        self._left_nuclide = left_nuclide
+
+    @right_nuclide.setter
+    def right_nuclide(self, right_nuclide):
+        cv.check_type('right_nuclide', right_nuclide,
+                      (openmc.Nuclide, CrossNuclide, AggregateNuclide))
+        self._right_nuclide = right_nuclide
+
+    @binary_op.setter
+    def binary_op(self, binary_op):
+        cv.check_type('binary_op', binary_op, str)
+        cv.check_value('binary_op', binary_op, _TALLY_ARITHMETIC_OPS)
+        self._binary_op = binary_op
+
+
+class CrossFilter(object):
+    """A special-purpose filter used to encapsulate all combinations of two
+    tally's filter bins as an outer product for tally arithmetic.
+
+    Parameters
+    ----------
+    left_filter : Filter or CrossFilter
+        The left filter in the outer product
+    right_filter : Filter or CrossFilter
+        The right filter in the outer product
+    binary_op : str
+        The tally arithmetic binary operator (e.g., '+', '-', etc.) used to
+        combine two tally's filter bins with this CrossFilter
+
+    Attributes
+    ----------
+    type : str
+        The type of the crossfilter (e.g., 'energy / energy')
+    left_filter : Filter or CrossFilter
+        The left filter in the outer product
+    right_filter : Filter or CrossFilter
+        The right filter in the outer product
+    binary_op : str
+        The tally arithmetic binary operator (e.g., '+', '-', etc.) used to
+        combine two tally's filter bins with this CrossFilter
+    bins : dict of Iterable
+        A dictionary of the bins from each filter keyed by the types of the
+        left / right filters
+    num_bins : Integral
+        The number of filter bins (always 1 if aggregate_filter is defined)
+
+    """
+
+    def __init__(self, left_filter=None, right_filter=None, binary_op=None):
+
+        left_type = left_filter.type
+        right_type = right_filter.type
+        self._type = '({0} {1} {2})'.format(left_type, binary_op, right_type)
+
+        self._bins = {}
+
+        self._left_filter = None
+        self._right_filter = None
+        self._binary_op = None
+
+        if left_filter is not None:
+            self.left_filter = left_filter
+            self._bins['left'] = left_filter.bins
+        if right_filter is not None:
+            self.right_filter = right_filter
```

Comment:
As a reader, storing duplicate data in a class in general feels dangerous. This is twice the amount of members/data that you have to keep track of, and most of all, keep in sync. If a new contributor comes in and does not realize there is duplicate data here to keep track of, they might only update one of the members of this class. A safer and more maintainable thing to do might be to add a helper function to simply get the left bin from the left_filter. Something like this
```
def left_bin(self):
    return self.left_filter.bins
```
This would then allow you to make a lot of the other code in this class more maintainable. For example, your `left_filter(self, left_filter)` setter would only one member assignment instead of two.